Support for Arduino MKRWifi 1010 (atmelsam)

Hi avaldebe, thank you for the great library.

I would like to use it on my MKRWifi 1010 board with Next-PM sensor , I tried to modify your PM serial Library and use it on my board but no luck. 
**If you can help me I will appreciate that.**
My platformio.ini looks like :  https://docs.platformio.org/en/latest/boards/atmelsam/mkrwifi1010.html

```ini
[env:mkrwifi1010]
platform = atmelsam
board = mkrwifi1010
```

My example:

```arduino
#include <PMserial.h>
SerialPM pms(NextPM, Serial1);  // PM Serial, RX, TX
void setup() {
  Serial.begin(9600);
 // Serial1.begin(9600, SERIAL_8N1, 13, 14);
  pms.init();                   // config serial port
}

void loop() {
  pms.read();                   // read the PM sensor
  Serial.print(F("PM1.0 "));Serial.print(pms.pm01);Serial.print(F(", "));
  Serial.print(F("PM2.5 "));Serial.print(pms.pm25);Serial.print(F(", "));
  Serial.print(F("PM10 ")) ;Serial.print(pms.pm10);Serial.println(F(" [ug/m3]"));
  delay(10000);                 // wait for 10 seconds
}
```

The errors:

```text
In file included from D:\Anass_PM\PMserial\PMserial.ino:2:0:
C:\Users\Asus\OneDrive\Documents\Arduino\libraries\PMserial_library1/PMserial.h:64:24: error: 'uart' has not been declared
   SerialPM(PMS sensor, uart &serial) : pms(sensor) {
                        ^~~~
C:\Users\Asus\OneDrive\Documents\Arduino\libraries\PMserial_library1/PMserial.h: In constructor 'SerialPM::SerialPM(PMS, int&)':
C:\Users\Asus\OneDrive\Documents\Arduino\libraries\PMserial_library1/PMserial.h:65:13: error: cannot convert 'int*' to 'Stream*' in assignment
     uart = &serial;
             ^~~~~~
D:\Anass_PM\PMserial\PMserial.ino: At global scope:
PMserial:7:29: error: cannot bind non-const lvalue reference of type 'int&' to an rvalue of type 'int'
 SerialPM pms(NextPM, Serial1);  // PM Serial, RX, TX
                             ^
In file included from C:\Users\Asus\OneDrive\Documents\ArduinoData\packages\arduino\hardware\samd\1.8.6\variants\mkrwifi1010/variant.h:165:0,
                 from C:\Users\Asus\OneDrive\Documents\ArduinoData\packages\arduino\hardware\samd\1.8.6\cores\arduino/delay.h:23,
                 from C:\Users\Asus\OneDrive\Documents\ArduinoData\packages\arduino\hardware\samd\1.8.6\cores\arduino/Arduino.h:81,
                 from sketch\PMserial.ino.cpp:1:
C:\Users\Asus\OneDrive\Documents\ArduinoData\packages\arduino\hardware\samd\1.8.6\cores\arduino/Uart.h:45:5: note:   after user-defined conversion: virtual Uart::operator bool()
     operator bool() { return true; }
     ^~~~~~~~
In file included from D:\Anass_PM\PMserial\PMserial.ino:2:0:
C:\Users\Asus\OneDrive\Documents\Arduino\libraries\PMserial_library1/PMserial.h:64:3: note:   initializing argument 2 of 'SerialPM::SerialPM(PMS, int&)'
   SerialPM(PMS sensor, uart &serial) : pms(sensor) {
   ^~~~~~~~
exit status 1
cannot bind non-const lvalue reference of type 'int&' to an rvalue of type 'int'
```

Thanks in advance.

--- a/.vscode/settings.json
+++ b/.vscode/settings.json
@@ -19,9 +19,11 @@
         "megaatmega",
         "mhetesp",
         "minikit",
+        "mkrwifi",
         "nbar",
         "nbin",
         "rhum",
+        "samd",
         "uint"
     ]
 }

--- a/.vscode/settings.json
+++ b/.vscode/settings.json
@@ -4,6 +4,7 @@
         "CKSUM",
         "HCHO",
         "HWSERIAL",
+        "Hzatmega",
         "MCU",
         "PMSA",
         "PMSerial",
@@ -11,6 +12,8 @@
         "SERIALPM",
         "UART",
         "USART",
+        "atmelavr",
+        "atmelsam",
         "bluepill",
         "chlen",
         "datasheet",
@@ -24,6 +27,8 @@
         "nbin",
         "rhum",
         "samd",
+        "stlink",
+        "ststm",
         "uint"
     ]
 }

--- a/README.md
+++ b/README.md
@@ -61,7 +61,7 @@ The sensor type inference does not cover the `PMS5003S` and `PMS5003T` variants,
 
 ### PMSx003 on HardwareSerial
 
-```arduino
+```c++
 #include <PMserial.h>
 SerialPM pms(PMSx003, Serial);  // PMSx003, UART
 
@@ -83,7 +83,7 @@ Setup for different MCU is covered on the [HardwareSerial][] example.
 
 ### PMSx003 on SoftwareSerial
 
-```arduino
+```c++
 #include <PMserial.h>
 SerialPM pms(PMSx003, 10, 11);  // PMSx003, RX, TX
 
@@ -108,30 +108,30 @@ Setup for different MCU is covered on the [SoftwareSerial][] example.
 On some ESP32 boards Serial1 default pins are connected to the flash.
 Using the standard constructor will cause a crash, see [espressif/arduino-esp32#148](https://github.com/espressif/arduino-esp32/issues/148).
 
-```Arduino
+```c++
 // will crash the ESP32
 SerialPM pms(PMSx003, Serial1);
-````
+```
 
 Fortunately, it is possible to define alternative for pins by calling:
 
-```Arduino
+```c++
 // define Serial1 pins
 Serial1.begin(9600, SERIAL_8N1, <RX>, <TX>);
 ```
 
 The PMSerial library uses this feature to implement the flexibility of SoftwareSerial
 
-```Arduino
+```c++
 // define Serial1 pins
 SerialPM pms(PMSx003, <RX>, <TX>);
-````
+```
 
 The [SoftwareSerial example][esp32hw] uses Serial1 on pins 23 (RX) and 19 (TX).
 The [HardwareSerial example][esp32hw] uses Serial2.
 
-[esp32hw]: examples/HardwareSerial/README.md#mhetesp32minikit
-[esp32sw]: examples/SoftwareSerial/README.md#mhetesp32minikit
+[esp32hw]: examples/HardwareSerial/README.md#esp32-minikit
+[esp32sw]: examples/SoftwareSerial/README.md#esp32-minikit
 
 ## Advanced usage
 
@@ -215,7 +215,7 @@ The available status/error codes and pre-defined error messages are:
 For easy of use, the error message are pre-defined with `#define`.
 See the [examples](examples/) for error handling implementation.
 
-## Contribute back
+## Contribute
 
 If you have read this far, this is the library for your project
 or you can not figure out how to use it. In any case, I can use your help.

--- a/README.md
+++ b/README.md
@@ -232,8 +232,9 @@ See issue [#4][GH4] for inspiration. PRs are welcomed.
 ## Changelog
 
 - Work in progress
-  - Support the PMS5003 S/ST/T variants [#10][GH10]
-  - Use Serial1 as "SoftwareSerial" for ESP32, [#7][GH7]
+  - Support `atmelsam` devices [#8][GH8]/[PR12][]
+  - Support the PMS5003 S/ST/T sensors [#10][GH10]
+  - Use Serial1 as "SoftwareSerial" for ESP32, [#7][GH7]/[PR9][]
 - 1.0.1
   - Fix broken SoftwareSerial for ESP8266, [#6][GH6]
   - ESP8266 use EspSoftwareSerial@>=6.7.1
@@ -242,4 +243,7 @@ See issue [#4][GH4] for inspiration. PRs are welcomed.
 
 [GH6]:  https://github.com/avaldebe/PMserial/issues/6
 [GH7]:  https://github.com/avaldebe/PMserial/issues/7
+[GH8]:  https://github.com/avaldebe/PMserial/issues/8
+[PR9]:  https://github.com/avaldebe/PMserial/pull/9
 [GH10]: https://github.com/avaldebe/PMserial/issues/10
+[PR12]: https://github.com/avaldebe/PMserial/pull/12

--- a/examples/HardwareSerial/README.md
+++ b/examples/HardwareSerial/README.md
@@ -2,12 +2,20 @@
 
 Read PMS5003 sensor on Serial/Serial1/Serial2, depending on the board.
 
-## Arduino `leonardo`
+## [Arduino Leonardo](https://docs.platformio.org/en/latest/boards/atmelavr/leonardo.html)
 
 **Note** This is a 5V board, and the PMS5003 RX/TX pins are 3.3V.
 You need to add a logic level converter, or use a 3.3V version of this board.
 
 - Serial1 on pins 0 (RX) and 1 (TX).
+
+```ini
+[env:leonardo]
+framework = arduino
+platform = atmelavr
+board = leonardo
+build_flags = -D USE_HWSERIAL1
+```
 
 ```bash
 # compile
@@ -20,7 +28,7 @@ platformio run -e leonardo -t upload
 platformio run -e leonardo -t monitor
 ```
 
-## Arduino `mega2560`
+## [Arduino Mega 2560](https://docs.platformio.org/en/latest/boards/atmelavr/megaatmega2560.html)
 
 **Note** This is a 5V board, and the PMS5003 RX/TX pins are 3.3V.
 You need to add a logic level converter, or use a 3.3V version of this board.
@@ -30,6 +38,14 @@ You need to add a logic level converter, or use a 3.3V version of this board.
 - Serial3 on pins 15 (RX) and 14 (TX).
 
 This example uses Serial1.
+
+```ini
+[env:mega2560]
+framework = arduino
+platform = atmelavr
+board = megaatmega2560
+build_flags = -D USE_HWSERIAL1
+```
 
 ```bash
 # compile
@@ -42,7 +58,7 @@ platformio run -e mega2560 -t upload
 platformio run -e mega2560 -t monitor
 ```
 
-## STM32 `genericSTM32F103C8`
+## [STM32F103C8](https://docs.platformio.org/en/latest/boards/ststm32/genericSTM32F103C8.html)
 
 **Note** This is a 3.3V board, and the PMS5003 requires 5V.
 You need provide 5V for the fan to operate properly.
@@ -52,6 +68,15 @@ You need provide 5V for the fan to operate properly.
 - Serial3 on pins PB11 (RX) and PB10 (TX).
 
 This example uses Serial1.
+
+```ini
+[env:genericSTM32F103C8]
+framework = arduino
+platform = ststm32
+board = genericSTM32F103C8
+upload_protocol = stlink
+build_flags = -D USE_HWSERIAL1
+```
 
 ```bash
 # compile
@@ -64,7 +89,7 @@ platformio run -e genericSTM32F103C8 -t upload
 platformio run -e genericSTM32F103C8 -t monitor
 ```
 
-## STM32 `maple_mini`
+## [STM32 Maple Mini](https://docs.platformio.org/en/latest/boards/ststm32/maple_mini_b20.html)
 
 **Note** This is a 3.3V board, and the PMS5003 requires 5V.
 You need provide 5V for the fan to operate properly.
@@ -74,6 +99,14 @@ You need provide 5V for the fan to operate properly.
 - Serial3 on pins 0/PB11 (RX) and 1/PB10 (TX).
 
 This example uses Serial2.
+
+```ini
+[env:maple_mini]
+framework = arduino
+platform = ststm32
+board = maple_mini_b20
+build_flags = -D USE_HWSERIAL2
+```
 
 ```bash
 # compile
@@ -86,7 +119,7 @@ platformio run -e maple_mini -t upload
 platformio run -e maple_mini -t monitor
 ```
 
-## ESP8266 `esp01`
+## [ESP8266 ESP-01 512k](https://docs.platformio.org/en/latest/boards/espressif8266/esp01.html)
 
 **Note** This is a 3.3V board, and the PMS5003 requires 5V.
 
@@ -95,6 +128,13 @@ platformio run -e maple_mini -t monitor
 This example used Serial for communicating with the sensor
 and printing the sensor values.
 Some garbage between printouts is to be expected.
+
+```ini
+[env:esp01]
+framework = arduino
+platform = espressif8266
+board = esp01
+```
 
 ```bash
 # compile
@@ -107,7 +147,7 @@ platformio run -e esp01 -t upload
 platformio run -e esp01 -t monitor
 ```
 
-## ESP32 `mhetesp32minikit`
+## [ESP32 MiniKit](https://docs.platformio.org/en/latest/boards/espressif32/mhetesp32minikit.html)
 
 **Note** This is a 3.3V board, and the PMS5003 requires 5V.
 You need provide 5V for the fan to operate properly.
@@ -118,29 +158,37 @@ You need provide 5V for the fan to operate properly.
 On some ESP32 boards Serial1 default pins are connected to the flash.
 Using the standard constructor will cause a crash, see [espressif/arduino-esp32#148](https://github.com/espressif/arduino-esp32/issues/148).
 
-```Arduino
+```c++
 // will crash the ESP32
 SerialPM pms(PMSx003, Serial1);
-````
+```
 
 Fortunately, it is possible to define alternative for pins by calling:
 
-```Arduino
+```c++
 // define Serial1 pins
 Serial1.begin(9600, SERIAL_8N1, <RX>, <TX>);
 ```
 
 The PMSerial library uses this feature to implement the flexibility of SoftwareSerial
 
-```Arduino
+```c++
 // define Serial1 pins
 SerialPM pms(PMS5003, <RX>, <TX>);
-````
+```
 
 The [SoftwareSerial example][esp32sw] uses Serial1 on pins 23 (RX) and 19 (TX).
 This example uses Serial2.
 
-[esp32sw]: ../SoftwareSerial/README.md#mhetesp32minikit
+[esp32sw]: ../SoftwareSerial/README.md#esp32-minikit
+
+```ini
+[env:esp32minikit]
+framework = arduino
+platform = espressif32
+board = mhetesp32minikit
+build_flags = -D USE_HWSERIAL2
+```
 
 ```bash
 # compile
@@ -153,12 +201,20 @@ platformio run -e esp32minikit -t upload
 platformio run -e esp32minikit -t monitor
 ```
 
-## Arduino `mkrwifi1010`
+## [Arduino MKR WiFi 1010](https://docs.platformio.org/en/latest/boards/atmelsam/mkrwifi1010.html)
 
 **Note** This is a 3.3V board, and the PMS5003 requires 5V.
 You need provide 5V for the fan to operate properly.
 
 - Serial1 on pins 13 (RX) and 14 (TX).
+
+```ini
+[env:mkrwifi1010]
+framework = arduino
+platform = atmelsam
+board = mkrwifi1010
+build_flags = -D USE_HWSERIAL1
+```
 
 ```bash
 # compile

--- a/examples/HardwareSerial/README.md
+++ b/examples/HardwareSerial/README.md
@@ -152,3 +152,21 @@ platformio run -e esp32minikit -t upload
 # open serial monitor
 platformio run -e esp32minikit -t monitor
 ```
+
+## Arduino `mkrwifi1010`
+
+**Note** This is a 3.3V board, and the PMS5003 requires 5V.
+You need provide 5V for the fan to operate properly.
+
+- Serial1 on pins 13 (RX) and 14 (TX).
+
+```bash
+# compile
+platformio run -e mkrwifi1010
+
+# upload
+platformio run -e mkrwifi1010 -t upload
+
+# open serial monitor
+platformio run -e mkrwifi1010 -t monitor
+```

--- a/examples/HardwareSerial/platformio.ini
+++ b/examples/HardwareSerial/platformio.ini
@@ -48,3 +48,8 @@ board = esp01
 platform = espressif32
 board = mhetesp32minikit
 build_flags = -D USE_HWSERIAL2
+
+[env:mkrwifi1010]
+platform = atmelsam
+board = mkrwifi1010
+build_flags = -D USE_HWSERIAL1

--- a/examples/HardwareSerial/platformio.ini
+++ b/examples/HardwareSerial/platformio.ini
@@ -44,7 +44,7 @@ build_flags = -D USE_HWSERIAL2
 platform = espressif8266
 board = esp01
 
-[env:mhetesp32minikit]
+[env:esp32minikit]
 platform = espressif32
 board = mhetesp32minikit
 build_flags = -D USE_HWSERIAL2

--- a/examples/SoftwareSerial/README.md
+++ b/examples/SoftwareSerial/README.md
@@ -2,12 +2,19 @@
 
 Read PMS5003 sensor on SWSerial
 
-## Arduino `uno`
+## [Arduino Uno](https://docs.platformio.org/en/latest/boards/atmelavr/uno.html)
 
 **Note** This is a 5V board, and the PMS5003 RX/TX pins are 3.3V.
 You need to add a logic level converter, or use a 3.3V version of this board.
 
 - SWSerial on pins 10 (RX) and 11 (TX).
+
+```ini
+[env:uno]
+framework = arduino
+platform = atmelavr
+board = uno
+```
 
 ```bash
 # compile
@@ -20,13 +27,20 @@ platformio run -e uno -t upload
 platformio run -e uno -t monitor
 ```
 
-## Arduino `mini168_3V3`
+## [Arduino Pro Mini (ATmega168, 3.3V, 8 MHz)](https://docs.platformio.org/en/latest/boards/atmelavr/pro8MHzatmega168.html)
 
 3.3V/8Mhz version of the pro mini board with an ATmega168.
 **Note** This is a 3.3V board, and the PMS5003 requires 5V.
 You need provide 5V for the fan to operate properly.
 
 - SWSerial on pins 10 (RX) and 11 (TX).
+
+```ini
+[env:mini168_3V3]
+framework = arduino
+platform = atmelavr
+board = pro8MHzatmega168
+```
 
 ```bash
 # compile
@@ -39,13 +53,20 @@ platformio run -e mini168_3V3 -t upload
 platformio run -e mini168_3V3 -t monitor
 ```
 
-## Arduino `mini328_3V3`
+## [Arduino Pro Mini (ATmega328, 3.3V, 8 MHz)](https://docs.platformio.org/en/latest/boards/atmelavr/pro8MHzatmega328.html)
 
 3.3V/8Mhz version of the pro mini board with an ATmega328.
 **Note** This is a 3.3V board, and the PMS5003 requires 5V.
 You need provide 5V for the fan to operate properly.
 
 - SWSerial on pins 10 (RX) and 11 (TX).
+
+```ini
+[env:mini328_3V3]
+framework = arduino
+platform = atmelavr
+board = pro8MHzatmega328
+```
 
 ```bash
 # compile
@@ -58,7 +79,7 @@ platformio run -e mini328_3V3 -t upload
 platformio run -e mini328_3V3 -t monitor
 ```
 
-## ESP8266 `esp01`
+## [ESP8266 ESP-01 512k](https://docs.platformio.org/en/latest/boards/espressif8266/esp01.html)
 
 **Note** This is a 3.3V board, and the PMS5003 requires 5V.
 
@@ -66,6 +87,15 @@ platformio run -e mini328_3V3 -t monitor
 - Needs [EspSoftwareSerial@>=6.7.1][issue6]
 
 [issue6]: https://github.com/avaldebe/PMserial/issues/6
+
+```ini
+[env:esp01]
+framework = arduino
+platform = espressif8266
+lib_deps = EspSoftwareSerial@>=6.7.1
+board = esp01
+build_flags = -D PMS_RX=2 -D PMS_TX=0
+```
 
 ```bash
 # compile
@@ -78,12 +108,21 @@ platformio run -e esp01 -t upload
 platformio run -e esp01 -t monitor
 ```
 
-## ESP8266 `d1_mini`
+## [WeMos D1 mini](https://docs.platformio.org/en/latest/boards/espressif8266/d1_mini.html)
 
 **Note** This is a 3.3V board, and the PMS5003 requires 5V.
 
 - SWSerial on pins GPIO13/D7 (RX) and GPIO12/D6 (TX).
 - Needs [EspSoftwareSerial@>=6.7.1][issue6]
+
+```ini
+[env:d1_mini]
+framework = arduino
+platform = espressif8266
+lib_deps = EspSoftwareSerial@>=6.7.1
+board = d1_mini
+build_flags = -D PMS_RX=D7 -D PMS_TX=D6
+```
 
 ```bash
 # compile
@@ -96,7 +135,7 @@ platformio run -e d1_mini -t upload
 platformio run -e d1_mini -t monitor
 ```
 
-## ESP32 `mhetesp32minikit`
+## [ESP32 MiniKit](https://docs.platformio.org/en/latest/boards/espressif32/mhetesp32minikit.html)
 
 **Note** This is a 3.3V board, and the PMS5003 requires 5V.
 You need provide 5V for the fan to operate properly.
@@ -107,29 +146,37 @@ You need provide 5V for the fan to operate properly.
 On some ESP32 boards Serial1 default pins are connected to the flash.
 Using the standard constructor will cause a crash, see [espressif/arduino-esp32#148](https://github.com/espressif/arduino-esp32/issues/148).
 
-```Arduino
+```c++
 // will crash the ESP32
 SerialPM pms(PMSx003, Serial1);
-````
+```
 
 Fortunately, it is possible to define alternative for pins by calling:
 
-```Arduino
+```c++
 // define Serial1 pins
 Serial1.begin(9600, SERIAL_8N1, <RX>, <TX>);
 ```
 
 The PMSerial library uses this feature to implement the flexibility of SoftwareSerial
 
-```Arduino
+```c++
 // define Serial1 pins
 SerialPM pms(PMS5003, <RX>, <TX>);
-````
+```
 
 This example uses Serial1 on pins 23 (RX) and 19 (TX).
 The [HardwareSerial example][esp32hw] uses Serial2 directly.
 
-[esp32hw]: ../HardwareSerial/README.md#mhetesp32minikit
+[esp32hw]: ../HardwareSerial/README.md#esp32-minikit
+
+```ini
+[env:esp32minikit]
+framework = arduino
+platform = espressif32
+board = mhetesp32minikit
+build_flags = -D PMS_RX=23 -D PMS_TX=19
+```
 
 ```bash
 # compile

--- a/examples/SoftwareSerial/platformio.ini
+++ b/examples/SoftwareSerial/platformio.ini
@@ -43,7 +43,7 @@ lib_deps = EspSoftwareSerial@>=6.7.1
 board = d1_mini
 build_flags = -D PMS_RX=D7 -D PMS_TX=D6
 
-[env:mhetesp32minikit]
+[env:esp32minikit]
 platform = espressif32
 board = mhetesp32minikit
 build_flags = -D PMS_RX=23 -D PMS_TX=19

--- a/examples/debug/README.md
+++ b/examples/debug/README.md
@@ -10,6 +10,7 @@ with calls to some special debugging methods.
 ## Boards tested
 
 - AVR: `uno`, `leonardo`, `megaatmega2560`
+- SAMD: `mkrwifi1010`
 - STM32: `bluepill_f103c8`, `maple_mini_b20`
 - ESP8266: `esp01`, `esp01_1m`, `d1_mini`
 - ESP32: `mhetesp32minikit`, `m5stack-core-esp32`

--- a/examples/debug/platformio.ini
+++ b/examples/debug/platformio.ini
@@ -76,3 +76,8 @@ build_flags = -D PMS_RX=23 -D PMS_TX=19
 ; esp32 (1.25Mb)
 platform = espressif32
 board = m5stack-core-esp32
+
+[env:mkrwifi1010]
+platform = atmelsam
+board = mkrwifi1010
+build_flags = -D USE_HWSERIAL1

